src/viztracer/modules/vcompressor/vc_dump.c

```diff
@@ -146,9 +147,296 @@ int dump_parsed_trace_events(PyObject* trace_events, FILE* fptr)
+
+    // read pid, tid, name and keys
+    READ_DATA(&pid, uint64_t, fptr);
+    READ_DATA(&tid, uint64_t, fptr);
+    name = freadstr(fptr);
+    READ_DATA(&arg_key_count, uint64_t, fptr);
+    counter_name = PyUnicode_FromString(name);
+    free(name);
+    counter_pid = PyLong_FromUnsignedLongLong(pid);
+    counter_tid = PyLong_FromUnsignedLongLong(tid);
+    for (uint64_t i = 0; i < arg_key_count; i++) {
+        freadstrn(buffer, STRING_BUFFER_SIZE - 1, fptr);
+        counter_arg_key = PyUnicode_FromString(buffer);
+        PyList_Append(arg_key_list, counter_arg_key);
+        Py_DECREF(counter_arg_key);
+    }
+
+    // read counter events
+    // cached_args stores the newest value
+    // current_arg stores the counter arg of current timestamp
+    READ_DATA(&counter_event_count, uint64_t, fptr);
+    for (uint64_t i = 0; i < counter_event_count; i++) {
+        current_arg = PyDict_New();
+        READ_DATA(&ts_64, uint64_t, fptr);
+        for (uint64_t j = 0; j < arg_key_count; j++) {
+            READ_DATA(&header, uint8_t, fptr);
+            counter_arg_key = PyList_GetItem(arg_key_list, j);
+            // counter arg not change means current value is same with it in last arg
+            // so we need to read it from cached_args 
+            // other state means current value is different from it in last arg
+            // so we need to read it from file and save it in cached_args
+            switch (header)
+            {
+                case VC_HEADER_COUNTER_ARG_NOT_KNOW:
+                    PyDict_SetItem(cached_args, counter_arg_key, Py_None);
+                    break;
+                case VC_HEADER_COUNTER_ARG_NOT_CHANGE:
+                    counter_arg_value = PyDict_GetItem(cached_args, counter_arg_key);
+                    if (counter_arg_value && counter_arg_value != Py_None) {
+                        PyDict_SetItem(current_arg, counter_arg_key, counter_arg_value);
+                    }
+                    break;
+                case VC_HEADER_COUNTER_ARG_LONG:
+                    READ_DATA(&value_longlong, int64_t, fptr);
+                    counter_arg_value = PyLong_FromLongLong(value_longlong);
+                    PyDict_SetItem(current_arg, counter_arg_key, counter_arg_value);
+                    PyDict_SetItem(cached_args, counter_arg_key, counter_arg_value);
+                    Py_DECREF(counter_arg_value);
+                    break;
+                case VC_HEADER_COUNTER_ARG_FLOAT:
+                    READ_DATA(&value_double, double, fptr);
+                    counter_arg_value = PyFloat_FromDouble(value_double);
+                    PyDict_SetItem(current_arg, counter_arg_key, counter_arg_value);
+                    PyDict_SetItem(cached_args, counter_arg_key, counter_arg_value);
+                    Py_DECREF(counter_arg_value);
+                    break;
+                case VC_HEADER_COUNTER_ARG_LONG_STRING:
+                    freadstrn(buffer, STRING_BUFFER_SIZE - 1, fptr);
```

Comment:
There is a non-ignorable possibility here that this integer representation is longer than `STRING_BUFFER_SIZE`. Maybe do a `freadstr` and free the memory here.